text/0009-multilink_command_handling.md

```diff
@@ -0,0 +1,45 @@
+  * Start date: 2018-05-21
+  * Contributors: Achermann Florian, <acfloria@ethz.ch>
+  * Related issues: 
+  
+# Summary
+
+The goal of this document is to present a design on how to handle commands which are sent over different links with significantly different latencies from the ground control station to a vehicle.
+  
+# Motivation
+
+When transmitting commands over multiple links with different latencies to a single vehicle the order in which the commands are sent might not be equal to the order in which they are received. This can lead to an undesired response of the vehicle especially when using commands such as MAV_CMD_DO_CHANGE_ALTITUDE or MAV_CMD_DO_REPOSITION.
+
+# Detailed Design
+
+The planned system requires changes in the QGC, Mavlink, and Autopilot Software. First in the Mavlink codebase a timestamp field is added to the following messages:
+
+  * COMMAND_INT
+  * COMMAND_LONG
+
+The timestamp field contains the vehicle time since the bootup. 
+
+In a seconds step the groundstation needs to be synchronized with the vehicle time. This can be achieved using the SYSTEM_TIME message. The groundstation tracks the system time for each vehicle and fills it in the timestamp field for every command which is sent. As long as no SYSTEM_TIME message is received the timestamp field is set to 0.
```

Comment:
This doesn't handle variations in link latency (which can be massive), and not in order delivery. I think a different solution is required.

Reply:
True, but the link latency could be estimated by a roundtrip of the SYSTEM_TIME message.

So the protocol could be changed in the following way:
1. GCS periodically estimates the link latency by a roundtrip of the SYSTEM_TIME message
2. Vehicle sends a separate SYSTEM_TIME to the GCS to synchronize GCS time and vehicle by considering the estimated link latency.

Reply:
After thinking about the timing a bit more I think it could also be possible to remove the synchronization and use the unix time. However, I am not sure if the unix time is available on all devices we are supporting.